// @sudoku/stores/keyboard.js
import { cursor } from './cursor';
import { gamePaused } from './game';
import { grid } from './grid';
import { derived, writable } from 'svelte/store';

// 原有的 keyboardDisabled
export const keyboardDisabled = derived(
	[cursor, grid, gamePaused],
	([$cursor, $grid, $gamePaused]) => {
		return ($cursor.x === null && $cursor.y === null) || $grid[$cursor.y][$cursor.x] !== 0 || $gamePaused;
	},
	false,
);

// --- 智能提示/回溯状态 ---

// 当前高亮的候选数字 (例如 [1, 5, 9])
export const hintCandidates = writable([]);

// --- 新增：已尝试过的数字 (用于回溯后标红) ---
export const triedCandidates = writable([]);

// 提示模式是否激活
export const isHintActive = writable(false);

// 监听光标移动：一旦用户移动光标，重置
cursor.subscribe(() => {
	resetKeyboardHints();
});

// 重置辅助函数
export function resetKeyboardHints() {
	isHintActive.set(false);
	hintCandidates.set([]);
	triedCandidates.set([]); // 清空已尝试列表
}

/**
 * 设置键盘提示
 * @param {number[]} candidates - 所有合法的候选数
 * @param {number[]} tried - (可选) 已经尝试过导致失败的数字
 */
export function setKeyboardHints(candidates, tried = []) {
	hintCandidates.set(candidates);
	triedCandidates.set(tried);
	isHintActive.set(true);
}
