// @sudoku/stores/keyboard.js
import { cursor } from './cursor';
import { gamePaused } from './game';
import { grid } from './grid';
import { derived, writable } from 'svelte/store';

/**
 * Derived store that encapsulates the logic for determining keyboard accessibility.
 * The keyboard state is set to disabled if:
 * 1. No cell is currently selected (cursor is null).
 * 2. The selected cell is a fixed digit (non-zero in the original grid).
 * 3. The game is currently paused.
 */
export const keyboardDisabled = derived(
	[cursor, grid, gamePaused],
	([$cursor, $grid, $gamePaused]) => {
		return ($cursor.x === null && $cursor.y === null) || $grid[$cursor.y][$cursor.x] !== 0 || $gamePaused;
	},
	false,
);

/** 
 * Collection of candidate digits recommended for highlighting (e.g., [1, 5, 9]). 
 */
export const hintCandidates = writable([]);

/** 
 * Collection of digits that have been previously attempted at the current decision point. 
 * Used for visual differentiation (e.g., marking tried numbers in a different color).
 */
export const triedCandidates = writable([]);

/** 
 * State flag indicating whether the hint/backtrack visualization mode is currently active. 
 */
export const isHintActive = writable(false);

/**
 * Observer for the cursor state.
 * Resets keyboard suggestions whenever the user relocates the focus to a different cell.
 */
cursor.subscribe(() => {
	resetKeyboardHints();
});

/**
 * Resets hint-related state variables to their default values.
 * Clears the active hint flag and empties the candidate and trial lists.
 */
export function resetKeyboardHints() {
	isHintActive.set(false);
	hintCandidates.set([]);
	triedCandidates.set([]);
}

/**
 * Configures the keyboard hint state with specified data.
 * 
 * @param {number[]} candidates - The set of valid candidate digits to highlight.
 * @param {number[]} tried - The set of digits already attempted in the current Speculative Branch.
 */
export function setKeyboardHints(candidates, tried = []) {
	hintCandidates.set(candidates);
	triedCandidates.set(tried);
	isHintActive.set(true);
}
