// @sudoku/sencode/puzzleParser.js
import InvalidPuzzleError from './errors/InvalidPuzzleError';
import SudokuWikiUrlDecoder from './decoders/SudokuWikiUrlDecoder';
import SencodeDecoder from './decoders/SencodeDecoder';


// ==========================================
// Decoder registry (a lightweight factory/registry).
// Stores decoders in order; first match wins.
// ==========================================
class DecoderRegistry {
    constructor() {
		/** @type {Array<{ canHandle: (input: string) => boolean, decode: (input: string) => any }>} */
		this.decoders = [];
	}

	/**
	 * @param {{ canHandle: (input: string) => boolean, decode: (input: string) => any }} decoder
	 */

    register(decoder) {
        this.decoders.push(decoder);
    }

	/**
	 * @param {string} input
	 * @returns {{ canHandle: (input: string) => boolean, decode: (input: string) => any } | null}
	 */

    find(input) {
        for (const decoder of this.decoders) {
            if (decoder && decoder.canHandle(input)) {
                return decoder;
            }
        }
        return null;
    }
}

const defaultRegistry = new DecoderRegistry();
defaultRegistry.register(new SudokuWikiUrlDecoder());
defaultRegistry.register(new SencodeDecoder());

const defaultParser = new (class PuzzleParserImpl {
    parse(input) {
        const decoder = defaultRegistry.find(input);
        if (!decoder) {
            throw new InvalidPuzzleError('Invalid custom Sudoku puzzle');
        }
        return decoder.decode(input);
    }
})();

export default class PuzzleParser {
	static parse(input) {
		return defaultParser.parse(input);
	}
}
