// @sudoku/history/actionHelpers.js
import { getTreeHistoryManager } from './historyManager.js';
import { modal } from '@sudoku/stores/modal';
import { userGrid } from '@sudoku/stores/grid';
import { cursor } from '@sudoku/stores/cursor';
import { candidates } from '@sudoku/stores/candidates';

const historyManager = getTreeHistoryManager();

// --- 新增：内部同步函数 ---
function syncHistoryToStore() {
  const currentNode = historyManager.getCurrentNode();
  if (!currentNode || !currentNode.state) return;

  const { board, cursor: cursorPos, candidates: candMap } = currentNode.state;

  // 1. 更新棋盘
  userGrid.set(JSON.parse(JSON.stringify(board)));

  // 2. 更新光标
  cursor.set(cursorPos.x, cursorPos.y);

  // 3. 更新笔记 (需要确保 candidates store 有 set 方法)
  if (candidates.set) {
    candidates.set(JSON.parse(JSON.stringify(candMap || {})));
  }
}

export function initGameHistory() {
  const result = historyManager.initGame('游戏开始');
  return result;
}

export function saveUserAction(cell, value, cellCandidates) {
  const backtrackable = cellCandidates && cellCandidates.length >= 2;
  
  let infos = null;
  if (backtrackable) {
    infos = {
      cell: { x: cell.x, y: cell.y },
      candidates: [...cellCandidates],
      tried: [value] 
    };
  }
  
  const description = backtrackable ? 
    `选择 (${cell.x}, ${cell.y}) = ${value} (可回溯)` :
    `选择 (${cell.x}, ${cell.y}) = ${value}`;

  const result = historyManager.saveState(description, backtrackable, infos);
  
  return result;
}

// --- 修改后的执行函数，加入了同步逻辑 ---

export function performBacktrack() {
  console.log('执行回溯...');
  const success = historyManager.backtrack();
  if (success) {
    syncHistoryToStore();
  } else if (modal && modal.show) {
    modal.show('backtrackInfo', {
      title: '无法回溯',
      message: '没有可回溯的节点或所有选项都已尝试'
    });
  }
  return success;
}

export function performUndo() {
  console.log('执行撤销...');
  const success = historyManager.undo();
  if (success) {
    syncHistoryToStore();
  } else if (modal && modal.show) {
    modal.show('undoInfo', {
      title: '无法撤销',
      message: '没有可撤销的操作'
    });
  }
  return success;
}

export function performRedo() {
  console.log('执行重做...');
  const success = historyManager.redo();
  if (success) {
    syncHistoryToStore();
  } else if (modal && modal.show) {
    modal.show('redoInfo', {
      title: '无法重做',
      message: '没有可重做的操作'
    });
  }
  return success;
}

export const history = {
  subscribe: (callback) => {
    const update = () => {
      const stats = historyManager.stats;
      const info = historyManager.getInfos();
      callback({
        ...stats,
        canBacktrack: stats.canBacktrack,
        backtrackInfo: info
      });
    };
    update();
    const interval = setInterval(update, 100);
    return () => {
      clearInterval(interval);
    };
  },
  
  getStats: () => historyManager.stats,
  getInfos: () => historyManager.getInfos(),
  canUndo: () => historyManager.stats.canUndo,
  canRedo: () => historyManager.stats.canRedo,
  canBacktrack: () => historyManager.stats.canBacktrack,
};
